doc/wallet/design/design.md

```diff
@@ -0,0 +1,93 @@
+* **libTx** - Library that provides lower-level transaction building, rangeproof and signing functions, highly-reusable by wallet implementors.
+* **Wallet Traits** - A set of generic traits defined within libWallet and the `keychain` crate . A wallet implementation such as Grin's current
+default only needs to implement these traits in order to provide a wallet:
+  * **WalletClient** - Defines communication between the wallet, a running grin node and/or other wallets
+  * **WalletBackend** - Defines the storage implementation of the wallet
+  * **KeyChain** - Defines key derivation operations
+  
+### Module-Specific Notes
+
+A full API-Description for each of these parts is still TBD (and should be generated by rustdoc rather than repeated here). However a few design
+notes on each module are worth mentioning here.
+
+##### Web Wallet Client / Static File Server
+
+This component is not a 3rd-party hosted 'Web Wallet' , but a client meant to be run on the local machine only by the wallet owner. It should provide
+a usable browser interface into the wallet, that should be functionally equivalent to using the command line but (hopefully) far easier to use.
+It is currently not being included by a default grin build, although the required listener is currently being run by default. To build and test this
+component, see instructions on the [project page](https://github.com/mimblewimble/grin-web-wallet). The 'Static File Server' is still under
+discussion, and concerns how to provide the web-wallet to the user in a default Grin build.
+
+##### Owner API / Foreign API
+
+The high-level wallet API has been split into two, to allow for different requirements on each. For instance, the Foreign API would listen on
+an external-facing port, and therefore potentially has different security requirements from the Owner API, which can simply be bound to localhost
+only.
+
+##### libTX
+
+Transactions are built using the concept of a 'Slate', which is a data structure that gets passed around to all participants in a transaction,
+with each appending their Inputs, Outputs or Signatures to it to build a completed wallet transaction. Although the current mode of operation in
+the default client only supports single-user - single recipient, an arbitrary number of participants to a transaction is supported within libTX.
+
+##### Wallet Traits
+
+In the current code, a Wallet implementation is just a combination of these three traits. The vast majority of functions within libwallet
+and libTX have a signature similar to the following:
+
+```
+pub fn retrieve_outputs<T: ?Sized, C, K>(
+!·wallet: &mut T,              
+!·show_spent: bool,            
+!·tx_id: Option<u32>,          
+) -> Result<Vec<OutputData>, Error>
+where     
+!·T: WalletBackend<C, K>,      
+!·C: WalletClient,
+!·K: Keychain,
+{  
+```
+
+With `T` in this instance being a class that implements  the `WalletBackend` trait, which is further parameterized with implementations of
```

Comment:
It seems that `WalletBackend` which is described above as

> **WalletBackend** - Defines the storage implementation of the wallet

is more than storage layer, but also let say controller which glue all parts together? Is it possible to just drop in some new storage method (like sqlite) and reuse the rest of functionality?

Reply:
It can be looked at that way, yes. It'll probably be worth factoring out the storage a bit more and creating another trait that's more of a piece of glue that contains all the other components, but not quite there yet. The basic abstraction is good enough for the high-level documentation at the moment, anyhow.